designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,218 @@
+	// this will default to one budget with a value to 10%. 
+	// +kubebuilder:validation:XValidation:message="'schedule' must be set with 'duration'",rule="!self.all(x, (has(x.schedule) && !has(x.duration)) || (!has(x.schedule) && has(x.duration)))" 
+	// +kubebuilder:default:={{nodes: "10%"}} 
+	// +kubebuilder:validation:MaxItems=50 
+	// +optional 
+	Budgets []Budget `json:"budgets,omitempty" hash:"ignore"` 
+}
+```
+
+#### Example 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    consolidationPolicy: WhenUnderutilized
+    genericPerActionControls:
+      Consolidation:
+        disruptAfter: "30m"
+        budgets:
+          - nodes: "20%"
+            schedule: "0 0 1 * *"
+            duration: "1h"
+      Drift:
+        disruptAfter: "1h"
+        budgets:
+          - nodes: "10%"
+            schedule: "0 0 * * 0"
+            duration: "2h"
+      Expiration:
+        disruptAfter: "Never"
+```
+
+#### Considerations 
+Some of the API choices for a given action seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each action. It appears there is a need for disruption controls within the budgets for each action, not just overall.
+
+This approach aligns well with controls that apply to all existing actions. The proposal presented here is similar to the one mentioned above in relation to the actions we allow to be defined (All, Consolidation, Drift, Expiration, Emptiness).
+
+This proposal is currently scoped for disruptionBudgets by action. However, we should also consider incorporating other generic disruption controls into the PerActionControls, even if we do not implement them immediately. Moving ConsolidateAfter and ExpireAfter into the per-action controls is a significant migration that requires careful planning and its own dedicated design. This proposal simply demonstrates a potential model that highlights the benefits of defining controls at a per-action level of granularity.
+
+### Pros + Cons 
+* 👍 This model starts to make more sense as we continue to add general behaviors that apply to all disruption actions where users will want control on the action level.
+* 👍 Provides place per action for generic controls intended to be shared across all disruption methods. While not all methods share the same disruption actions, there are already cases for this. 
+* 👍 Could extend other fields beyond generic values for specific actions with validation   
+*  👎👎 Breaking API Change for Budgets at least, and if we decide to model DisruptAfter we also would have to break those apis. It might make sense to break the budgets now before they have garnered large adoption as it becomes harder to make the change as time goes on.
+* 👎 Doesn't allow for easy defaulting of `All` disruption actions
```

Comment:
This is a big negative to me, personally. Not being able to simply specify that I want a budget to apply to everything means that I have to go through the effort of specifying an `All` block, which feels odd to me

Reply:
Yes i agree, its really awkward, One shouldn't be forced to include the action. One of my reasons for leaning away from this option :(

Reply:
Yea, I think that All will be the main path for a lot. "just don't disrupt me here", or "allow for x number of machines at a time" because I want changes faster.